app/models/katello/concerns/pulp_database_unit.rb

```diff
@@ -119,71 +124,87 @@ def import_for_repository(repository)
               model.repository_id = repository.id unless many_repository_associations
               service.update_model(model)
             end
-            pulp_ids << pulp_id
+            pulp_id_href_map[pulp_id] = backend_identifier
           end
         end
-        sync_repository_associations(repository, :pulp_ids => pulp_ids) if self.many_repository_associations
+        sync_repository_associations(repository, :pulp_id_href_map => pulp_id_href_map) if self.many_repository_associations
       end
 
       def sync_repository_associations(repository, options = {})
+        associated_ids, id_href_map_for_repository = [], {}
         additive = options.fetch(:additive, false)
-        pulp_ids = options.fetch(:pulp_ids, nil)
-        associated_ids = with_pulp_id(pulp_ids).pluck(:id) if pulp_ids
-
-        table_name = self.repository_association_class.table_name
-        attribute_name = unit_id_field
+        pulp_id_href_map = options.dig(:pulp_id_href_map) || {}
+        pulp_ids = options.dig(:pulp_ids) || pulp_id_href_map.try(:keys)
+        ids_for_repository = with_pulp_id(pulp_ids).pluck(:id, :pulp_id)
+        associated_ids = ids_for_repository.map(&:first)
+        id_href_map_for_repository = Hash[*ids_for_repository.flatten]
+        id_href_map_for_repository.each_pair { |k, v| id_href_map_for_repository[k] = pulp_id_href_map[v] }
 
         existing_ids = self.repository_association_class.uncached do
-          self.repository_association_class.where(:repository_id => repository).pluck(attribute_name)
+          self.repository_association_class.where(:repository_id => repository).pluck(unit_id_field)
         end
-        new_ids = associated_ids - existing_ids
-        delete_ids = existing_ids - associated_ids
 
-        queries = []
-
-        if delete_ids.any? && !additive
-          queries << "DELETE FROM #{table_name} WHERE repository_id=#{repository.id} AND #{attribute_name} IN (#{delete_ids.join(', ')})"
+        if !additive && (delete_ids = existing_ids - associated_ids).any?
+          query = "DELETE FROM #{self.repository_association_class.table_name} WHERE repository_id=#{repository.id} AND #{unit_id_field} IN (#{delete_ids.join(', ')})"
+          ActiveRecord::Base.connection.execute(query)
         end
 
-        unless new_ids.empty?
-          inserts = new_ids.map { |unit_id| "(#{unit_id.to_i}, #{repository.id.to_i}, '#{Time.now.utc.to_s(:db)}', '#{Time.now.utc.to_s(:db)}')" }
-          queries << "INSERT INTO #{table_name} (#{attribute_name}, repository_id, created_at, updated_at) VALUES #{inserts.join(', ')}"
-        end
-
-        ActiveRecord::Base.transaction do
-          queries.each do |query|
-            ActiveRecord::Base.connection.execute(query)
-          end
+        unless (new_ids = associated_ids - existing_ids).empty?
+          self.repository_association_class.import(db_columns_sync, db_values(new_ids, id_href_map_for_repository, repository), validate: false)
```

Comment:
I realized that the delete and imports are happening in two separate transactions now. Before they were happening together. Should we wrap these in a single transaction?

Reply:
Looking at this, do you think wrapping both conditional blocks in a transaction would make sense or is that too big of a block?
```
ActiveRecord::Base.transaction do
  if !additive && (delete_ids = existing_ids - associated_ids).any?
    query = "DELETE FROM #{self.repository_association_class.table_name} WHERE repository_id=#{repository.id} AND #{unit_id_field} IN (#{delete_ids.join(', ')})"
    ActiveRecord::Base.connection.execute(query)
  end
  unless (new_ids = associated_ids - existing_ids).empty?
    self.repository_association_class.import(db_columns_sync, db_values(new_ids, id_href_map_for_repository, repository), validate: false)
  end
end
```

Reply:
Should be quick operations so I am OK with wrapping those in the transaction block